src/notification.c

```diff
@@ -234,129 +234,184 @@ check_value_type (const char *key,
 
-  if (!has_action)
+      g_variant_builder_add (builder, "{sv}", "label", label);
+      g_variant_builder_add (builder, "{sv}", "action", action);
+      if (target)
+        g_variant_builder_add (builder, "{sv}", "target", target);
+
+      g_variant_builder_close (builder);
+    }
+  else
     {
-      g_set_error_literal (error,
-                           XDG_DESKTOP_PORTAL_ERROR,
-                           XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                           "action key is missing");
-      return FALSE;
+      g_debug ("Notification button needs a label and action property");
     }
 
+
   return TRUE;
 }
 
 static gboolean
-check_buttons (GVariant *value,
-               GError **error)
+parse_buttons (GVariantBuilder  *builder,
+               GVariant         *value,
+               GError          **error)
 {
+  gboolean result = TRUE;
   int i;
 
   if (!check_value_type ("buttons", value, G_VARIANT_TYPE ("aa{sv}"), error))
     return FALSE;
 
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("{sv}"));
+  g_variant_builder_add (builder, "s", "buttons");
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("v"));
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("aa{sv}"));
+
   for (i = 0; i < g_variant_n_children (value); i++)
     {
       g_autoptr(GVariant) button = g_variant_get_child_value (value, i);
 
-      if (!check_button (button, error))
+      if (!parse_button (builder, button, error))
         {
           g_prefix_error (error, "invalid button: ");
-          return FALSE;
+          result = FALSE;
+          break;
         }
     }
-  return TRUE;
+
+  g_variant_builder_close (builder);
+  g_variant_builder_close (builder);
+  g_variant_builder_close (builder);
+
+  return result;
 }
 
 static gboolean
-check_serialized_icon (GVariant *value,
-                       GError **error)
+parse_serialized_icon (GVariantBuilder  *builder,
+                       GVariant         *icon,
+                       GError          **error)
 {
-  g_autoptr(GIcon) icon = g_icon_deserialize (value);
+  const char *key;
+  g_autoptr(GVariant) value = NULL;
 
-  if (!icon)
+  /* Since the specs allow a single string as icon name we need to keep support for it */
+  if (g_variant_is_of_type (icon, G_VARIANT_TYPE_STRING))
     {
-      g_set_error_literal (error,
-                           XDG_DESKTOP_PORTAL_ERROR,
-                           XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                           "invalid icon");
-      return FALSE;
+      g_autoptr(GIcon) deserialized_icon = NULL;
+
+      deserialized_icon = g_icon_deserialize (icon);
+
+      if (G_IS_THEMED_ICON (deserialized_icon))
+        {
+          g_autoptr(GVariant) serialized_icon = NULL;
+
+          serialized_icon = g_icon_serialize (deserialized_icon);
+
+          g_variant_builder_add (builder, "{sv}", "icon", serialized_icon);
+          return TRUE;
+        }
```

Comment:
this falls through to `check_value_type` and then fails there with a bit misleading error. might want to return here and get the error message right.